### PR TITLE
Use database as source of thruth

### PR DESCRIPTION
**Pull request by SaraVieira**
This PR does a bunch but it focus on two things:

- Move from  https://github.com/DamianOsipiuk/vue-query to https://tanstack.com/query/v4/docs because the second is more heavily maintained and has more community support
- Remove the idea of local state
	We used to save local state on changes and send that to the server but when using vue query we can update our UI by only using the data we have in the database by taking advantage of the cache and this cleans the code a whole bunch

### CODE REVIEW

**Comment by SaraVieira** [comment]
CleanShot 2022-11-30 at 16 32 49@2x

Love to see it

File: composables/useSingleReport.js [hunk 1]
```diff
@@ -1,35 +1,73 @@
...
+        method: "PUT",
+        body: {
+          ...report.value,
+          [pkg]: {
+            ...report.value?.state?.[pkg],
+            [eula]: report.value?.state?.[pkg]?.[eula] ? false : true,
+          },
+        },
+      }),
+    onSuccess: () => {
```
Comment (sarahghp) [inline composables/useSingleReport.js]: **question:** Are we handling these errors?

**suggestion, non-blocking, but followup needed:** Let's add some error handling.

File: components/AddANote.vue [hunk 2]
```diff
@@ -1,55 +1,46 @@
...
 
-/*
-  Right now this just updates state —
-  we should make it actually save and give feedback
-*/
-const saveNote = () => {
+const saveNote = async () => {
   buttonText.value = "...";
-  addNote({ pkg: props.name, note: tempFormValue.value });
+  await addNote({ pkg: props.name, note: tempFormValue.value });
```
Comment (sarahghp) [inline components/AddANote.vue:30]: **suggestion:** If we keep the two values as noted above, we should be able to do the update here with the return now that you (**praise**) updated things so that we don't need to pass down the prop.

File: components/ReportSwitch.vue [hunk 3]
```diff
@@ -17,21 +17,23 @@ const props = defineProps({
...
-const toggled = ref(currentlyActive(props.criterion));
 const onToggleChange = () => {
   toggleEulaPackageApproval({ pkg: props.name, eula: props.criterion });
 };
 </script>
 
 <template>
   <AxoSwitch
-    v-model:toggled="toggled"
+    :key="props.name"
```
Comment (sarahghp) [inline components/ReportSwitch.vue:35]: **praise:** Thank you, I always forget this.

**Review by sarahghp — COMMENTED** [review]
Thanks for doing all of this, @SaraVieira. 

I know it looks like there are a lot of comments — but I promise they are not terrible and most are small! 

Part of it is that there are a lot of changes to different parts of the system. Also the notes cover a few different things: notes to us about the projects, some big UX questions, and some small nitpicks that should go fast.

The two largest things are I think the state changes have some knock-on effects and I think we are missing some error handling. The second can be handled in a follow-up or here, as you see fit, but I think they are part of the definition of done for me. Please tell me if that seems off.

Also, I pushed a Vite config update. It makes it so we can run locally using `yarn link` with Fringe. They are grumpy about file system access for security reasons (fair enough).

Last, for my notes, I used [conventional comments](https://conventionalcomments.org/), which is a way of being clearer about one's intentions, developed by one of my former coworkers. I think it can be helpful sometimes, but let me know if it is annoying, and I will stop. :)

File: components/AddANote.vue [hunk 4]
```diff
@@ -1,55 +1,41 @@
...
+
   setTimeout(() => (buttonText.value = "+"), 500);
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply components/AddANote.vue:32]: truuuuuu, done!

File: composables/useSingleReport.js [hunk 5]
```diff
@@ -19,11 +19,10 @@ export const useSingleReport = () => {
     retry: 0,
   });
 
-  const areAllEulasApproved = ({ name, suggested_criteria }) => {
-    return suggested_criteria.every(
+  const areAllEulasApproved = ({ name, suggested_criteria }) =>
+    suggested_criteria.every(
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply composables/useSingleReport.js:23]: Honestly super okay with both so I changed it :)

File: components/ReportList.vue [hunk 6]
```diff
@@ -1,37 +1,18 @@
...
+import { watchEffect } from "vue";
 import {
   RadioGroup,
   RadioGroupLabel,
   RadioGroupOption,
   RadioGroupDescription,
 } from "@headlessui/vue";
-import { useSingleReport } from "../composables";
+import { useSingleReport, useCommit } from "../composables";
 import ShieldIcon from "./Icons/ShieldIcon.vue";
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply components/ReportList.vue:10]: Made an issue for that 

https://github.com/axodotdev/harbor/issues/39

File: components/ReportList.vue [hunk 7]
```diff
@@ -42,7 +23,7 @@ const selected = useState(() =>
     : report.value.suggestions[0]
 );
 
-onMounted(async () => {
+watchEffect(async () => {
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply components/ReportList.vue:26]: Also only found out yesterday, so much cleaner!

File: components/ReportList.vue [hunk 8]
```diff
@@ -9,20 +9,27 @@ import {
...
+    $fetch(`/api/reports/${params.id}/commit`, {
+      method: "POST",
+    }),
+  onSuccess: () =>
+    createToast("Commit triggered!", {
+      type: "success",
+      hideProgressBar: true,
+    }),
+  onError: () =>
+    createToast("There has been an error", {
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply components/ReportList.vue]: The issue here is that would also require backend work, im gonna make an issue to make sure our errors are good for people

[the remark this answers is not in the document]
Comment (SaraVieira) [reply components/ReportList.vue]: Issue created: https://github.com/axodotdev/harbor/issues/40

File: components/SingleReport.vue [hunk 9]
```diff
@@ -36,10 +36,10 @@ const createLabelsFromCriteria = (currentCriterion) => {
   <div class="flex flex-col justify-between h-full mt-12">
     <div class="py-6 px-4 sm:px-6 lg:px-8">
       <div>
-        <div class="flex justify-between items-center">
-          <h1 class="mb-0">{{ props.report.name }}</h1>
+        <div class="flex justify-between items-center gap-6">
+          <h1 class="mb-0 break-all">{{ props.report.name }}</h1>
           <AxoLink
-            class="bg-axo-orange text-sm md:text-base block text-center p-2 rounded shadow"
+            class="bg-axo-orange text-sm md:text-base block text-center p-2 rounded shadow min-w-[250px]"
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply components/SingleReport.vue:42]: oh, didnt know about that class, changed!

File: composables/useSingleReport.js [hunk 10]
```diff
@@ -46,7 +46,7 @@ export const useSingleReport = () => {
       $fetch(`/api/reports/${route.params.id}`, {
         method: "PUT",
         body: {
-          ...report.value,
+          ...report.value.state,
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply composables/useSingleReport.js:48]: The difference is because we used to use the packageState (https://github.com/axodotdev/harbor/pull/38/files#diff-73c4038e65f57f315724f998e9e12b6a85fa911c104b03824aacab85b2e1f010L11) and that didn't save everything but only the `report.state` and now we just have everything in one place

File: composables/useSingleReport.js (hunk 1 above)
Comment (SaraVieira) [reply composables/useSingleReport.js]: Wasn't in the aproval toggle, fixed

File: components/ReportSwitch.vue [hunk 11]
```diff
@@ -18,17 +18,22 @@ const props = defineProps({
...
 });
 const { report, toggleEulaPackageApproval } = useSingleReport();
-const currentlyActive = (eula) => report.value.state?.[props.name]?.[eula];
 
-const toggled = ref(currentlyActive(props.criterion));
+const toggled = ref(report.value.state?.[props.name]?.[props.criterion]);
+
+watchEffect(() => {
+  toggled.value = report.value.state?.[props.name]?.[props.criterion];
+});
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply components/ReportSwitch.vue]: I cleaned up to use v-model again, let me know your thoughts

File: components/AddANote.vue (hunk 2 above)
Comment (SaraVieira) [reply components/AddANote.vue]: Replied in nabove message

File: components/AddANote.vue [hunk 12]
```diff
@@ -1,55 +1,46 @@
...
 });
 
 const tempFormValue = ref("");
-const { addNote } = usePackageState();
+const { addNote, report } = useSingleReport();
 
-const noteInnerText = computed(() => {
-  return tempFormValue.value || props.defaultNote;
+watchEffect(() => {
+  tempFormValue.value = report.value?.state?.[props.name]?.note;
```
[the remark this answers is not in the document]
Comment (SaraVieira) [reply components/AddANote.vue]: hmmm, I made some changes to this file to make it work with the v-model and also try to clean it up. would love your opinion on the changes


It adds the two values back

[the remark this answers is not in the document]
Comment (sarahghp) [reply components/AddANote.vue]: Thanks! Unfortunately, this reintroduces the bug fixed in https://github.com/axodotdev/harbor/pull/33. If you add a note in one report and navigate to another report, the note box will show the note from the first report. That's why the `addNote` function was re-setting the temp variable at the end. 

(I wrote about it in more detail, although using the options API in https://github.com/axodotdev/harbor/pull/34.)

note-persists

I think to fix this, we'd probably have the local `addNote` function reset the temp value as before, but if we are interested in the simplification I mention above, we could put this off.

File: components/ReportSwitch.vue (hunk 11 above)
[the remark this answers is not in the document]
Comment (sarahghp) [reply components/ReportSwitch.vue]: This change throws a warning, unfortunately ... 

Screen Shot 2022-12-02 at 2 41 51 PM

...and that warning opened a little saga.

At first I thought it appeared just because we were using a `computed` value for a `v-model`, which we should avoid because they don't have setters and conceptually are reserved for reactive values, with `data` or `ref` reserved for values we change. 

However, on actually reading the error, I see it is because the value we are using for `report` is being returned from `useQuery` and isn't writable for whatever reason is internal to the library. If we fix the error, by severing the object that is returned as `report` in `components/ReportSwitch.vue`, then the rest of the values in `composables/useSingleReport.js` that rely on `report` being globally changed will fail. 

When I try to sever the value from that being returned by `useQuery` in `composables/useSingleReport.js` (by doing something like `const report = Object.assign({}, data);` it creates an error that I don't fully understand — nothing is returned. 

I also notice that we are getting at least a one-second or so lag in updating the list when we toggle now that does not exist on `main`:

lag

I believe that is because we are waiting for the toggle to write to the database before updating the list, which makes a bit of confusing UX. I think the ideal solution for that would be to 

All of this is to say, it appears to me that the one solution would be to:

1. Change `computed` to `ref`
2. Sever the object in  `components/ReportSwitch.vue`
3. Change the `toggleEulaPackageApproval` to take the entire report as an argument
4. Create a new issue for the timing, which based on what we have would likely need to be changed by emitting an event and then re-set if the DB failed, which would be a bit complex here.

A different opportunity here would be to radically simplify the way we are using state and updating the database. I've been thinking about this a little bit and originally was thinking it would be better to approach later (if at all), but I think maybe it would save us time in the end?

For that we would go old school and:
- create a shared data-structure in the parent component that feeds both the list and the detail view
- update it using events (Vue's `emit` is so nice for this) and then trigger the database calls at the top
  - basically update data structure optimistically, run db call, reset values on error
- pass the relevant data down as props

Fundamentally, this is a pretty simple app and I think a restructure might save us a lot of time and make it easier to accomplish the things we would want to with error handling, etc. If it ever got so large we needed more complex data-handling, we would then have a clearer idea of what needed that kind of handling. What do you think? Too much time? Worth it in the long run?

**Review by sarahghp — COMMENTED** [review]
Small notes related to the bigger comments above.

**Comment by SaraVieira** [comment]
@sarahghp 

About the AddANote:

That makes sense it does explain the reason for that increment, I managed to find a good solution I think, I changed the `tempFormValue` to be an Object that holds all the temporary values depending on what package we are in and this way when changing package the value will be cleaned and when going back we also do not loose what we were typing

How do you feel about this?
